### ISSUE TITLE
[Feature Request] Referencing equations, figures, sections, etc.

### ISSUE THREAD

**Issue by Wieke** [Zettlr/Zettlr#248]
It would be really neat if we could reference equations, figures, sections and such in a way similar to how latex does it. (Attach a label to a thing then reference the label and let the typesetter figure out what number exact number thing it is.) As someone who tends to liberally sprinkle equations throughout their writing this kind of feature is a must-have.

So I googled for 5 seconds and found the `pandoc-crossref` filter which, when combined with `pandoc-citeproc`, does exactly that. According to the [docs](http://lierdakil.github.io/pandoc-crossref/) allows for referencing equations, tables, figures, sections, codeblocks, and so forth. After installing `pandoc-crossref` and quck hack of `zettlr-export.js` I got Zettlr to export his:

```markdown
# Section reference test

Section @sec:test is a section.
 
## Equation reference test {#sec:test}

Eq @eq:test is an equation.

$$ 
x = y 
$$ {#eq:test}
```
as this:

![image](https://user-images.githubusercontent.com/6737186/60026666-45b40100-969c-11e9-9007-0b1704f5190b.png)

So that seems to work. Some considerations:

 * The reference autocomplete (now only used for citations) should probably be expanded with the labels in the document that is being edited.
 * Highlighting should probably be changed to de-emphasize the label definitions. (It's a bit silly to see a massive bold `{#sec:test}` behind the section header.)
 * The filter `pandoc-crossreff` was a bit of a pain to install (had to build it from source cause the archlinux package appears to be broken) so it might be worthwhile to have zettlr check if the filter exists and only use it if it does. That way the filter is an optional dependency. (I don't know how bad it is on other platforms.)
 * Personally I don't like the prefixes (`sec.` and `eq.`) it adds automatically, this is configurable in the filter.

**Comment by nathanlesage** [comment]
What do you mean by quick hack? Basically, if it's an optional thing it is certainly worth considering. Question is, is there any possibility this could interfere with other libraries or functions of Zettlr?

**Comment by Wieke** [comment]
All I did was add `--filter pandoc-crossref` to `_citeprocOptions` in [this line](https://github.com/Zettlr/Zettlr/blob/929c4d22ca3f3e89856d1b38767561c3b6131a75/source/main/zettlr-export.js#L129), it's not how you'd probably want to do it for real hence I called it a quick hack.

I can't say for certain, since I don't know all the ins and outs of Zettlr, but I'd be really surprised if it conflicts with anything. It's a pandoc filter, so it only comes into play when exporting, and it's intended to be used with `pandoc-citeproc` the only other pandoc filter zettlr already uses. The only way I could see it interfering is if the tag syntax is already being used for something, but as far as I know that isn't the case..

**Comment by nathanlesage** [comment]
> All I did was add --filter pandoc-crossref to _citeprocOptions in this line

Well, you could've easily just added the `--filter pandoc-crossref` to the pandoc command in the settings; would've spared you recompiling =D

But I see, I'll keep it on my agenda!

**Comment by Wieke** [comment]
Discovered a minor interference, the labels mess up the recognition of multiline equations.

**Comment by nathanlesage** [comment]
Oh that is good to know. But I think I see where this comes from (the last line does not only contain $$, but also the label).

**Comment by kastork** [comment]
I haven't tried figures and equations, but adding --filter pandoc-crossref to the pandoc command does not work for citations.  At least it doesn't cause citations to become links to the reference list entry, which is what I would have expected. 

If you add `--filter pandoc-crossref` right before `$citeproc$` in the command line definition, there is no effect.  If you add it right after `$citeproc`, the reference is not processed at all and you get `@whatever`

Zettlr 1.6.0
I'm on a mac, I installed pandoc, pandoc-citeproc and pandoc-crossref using homebrew

**Comment by Wieke** [comment]
@kastork If you want citations to be links you can add `link-citations: true` to the yaml frontmatter. 

And yeah crossref needs to be added before citeproc because they use the same syntax, it says so in the [crossref docs](http://lierdakil.github.io/pandoc-crossref/#pandoc-citeproc-and-pandoc-crossref).

**Comment by nathanlesage** [comment]
> crossref needs to be added before citeproc

This sounds like something for the Settings- and FAQ-page of the docs, right? If so, please feel free to PR! :)

**Comment by Wieke** [comment]
Well currently crossref isn't used in zettlr by default, nor is it mentioned anywhere in the docs. I'd assume that anybody who changes the pandoc command in the advanced settings knows to reference the documentation of pandoc or the pandoc filter they're using.

Though I have been thinking of writing a little retrospective thingy on what it was like to write my master thesis using zettlr (once i actually finish writing it). I've had to figure some things out (like the crossref thing) and it could be nice to lay it all out.

**Comment by nathanlesage** [comment]
Sounds pretty good! If you've done so, let me know, then we can make this public and let others know about it!

**Comment by tobiasdiez** [comment]
I think this would be a valuable addition (especially for mathematical-oriented people). @nathanlesage can you assign the issue to me?

**Comment by nathanlesage** [comment]
> I think this would be a valuable addition (especially for mathematical-oriented people). @nathanlesage can you assign the issue to me?

Nope, I can somehow only assign things to me, I think I need to add some people as members to this repository, which would also give some more privileges to more people … I'll have a look!

**Comment by nathanlesage** [comment]
@tobiasdiez I just sent you an invitation to a team with much more permissions so that you should be able to manage all this stuff here!

**Comment by mcaruel** [comment]
> > All I did was add --filter pandoc-crossref to _citeprocOptions in this line
> 
> Well, you could've easily just added the `--filter pandoc-crossref` to the pandoc command in the settings; would've spared you recompiling =D
> 
> But I see, I'll keep it on my agenda!

Hi I would like to do cross-referencing. It seems that the above post provides a solution, but I don't know how to implement it. Where exactly should the `filter Pandoc-crossref` be added if I don't want to build the program from the source. Where in the preference (in the asset manager ?) do I have to go and what exactly should I write ? Thanks for your help. Being able to cross-reference is very important for scientific writing, specially when there are a lot of formula.

**Comment by piccolbo** [comment]
One difference with Latex is that one has to point to a section. You can't just sprinkle labels closest to the related text. One  example is  that if I have

```
# section {#sec:this-section}

concept one and concept two
```

and elsewhere I want to reference each of the concepts, I insert `@sec:this-section`. If I later rearrange the section into two, one per concept, the references will become incorrect, silently and away from where I  am actually editing. In latex I would write

```
concept one \label{concept-one} and concept two \label{concept-two}
```
and as long as I keep the labels near the concepts, the numbering will be updated just fine. This left me with three options: complete my book without any references and then add all of them without any further reorg at the section level (numbering can never change again); accept a large number of incorrect references in the final version; write my own filter to match the latex behavior more closely. I went with number 3, unfortunately it's hacky enough that I don't feel like publishing it. But if you want a latex-like functionality in zettlr, I am afraid pandoc-crossref is quite a ways.

**Comment by Martin15135215** [comment]
Hi all, when will this cross-ref be available in Zettlr?

**Comment by victormnielsen** [comment]
I second this feature request

**Comment by carljoseph** [comment]
I third'll this feature request! 

I tried adding `- pandoc-crossref` to the list of filters in the export settings, but get this error when I run an export:

Export failed: WARNING: pandoc-crossref was compiled with pandoc 2.18 but is being run through 3.1.13. This is not supported. Strange things may (and likely will) happen silently. pandoc-crossref: Error in $: Incompatible API versions: encoded with [1,23,1] but attempted to decode with [1,22,2]. CallStack (from HasCallStack): error, called at src/Text/Pandoc/JSON.hs:112:48 in pndc-typs-1.22.2-031ed536:Text.Pandoc.JSON Error running filter pandoc-crossref: Filter returned error status 1

**Comment by nathanlesage** [comment]
So, adding crossref as a filter is now very easy, and I've just also implemented support for pre-rendering those cross-references.

To @carljoseph: I hope you were able to deduce what the error message tells you: Pandoc-crossref must match the Pandoc you use. So if you use Pandoc 3.1.13, you will need to use pandoc-crossref 3.1.13, too. It's a bit weird, but that's how it works.

**Comment by jgsw** [comment]
Brilliant that this is now fixed. If pandoc-crossref needs to be updated every time Pandoc is updated, then would it make sense (as an enhancement request) to bundle pandoc-crossref with Zettlr? That way, Zettlr's Pandoc-crossref version would always match Zettlr's Pandoc version.

**Comment by nathanlesage** [comment]
Hmhm, it's just ... the installer is already so big, and many people at least on macOS are using homebrew too, and actually don't need the bundled Pandoc at all... I'm currently thinking of maybe not shipping Pandoc and instead implementing a GUI to download that thing and keeping it updated if needed, but that will take time obviously!
